File: python/array_mean.py
# Mean of Array using Iteration and Recursion
from common import tramp
import itertools
import functools


def array_mean(arr: list) -> float:
    sum = 0
    for el in arr:
        sum += el
    return sum/len(arr)


def array_mean_rec(arr: list) -> float:
    length = len(arr)

    def helper(idx: int):
        if idx >= length:
            return 0
        return arr[idx]+helper(idx+1)
    return helper(0)/length


def array_mean_tail_rec(arr: list, acc=0, total_nums=0) -> float:
    if not arr:
        return acc/total_nums
    curr = arr[0]
    return array_mean_tail_rec(arr[1:], acc+curr, total_nums+1)


def array_mean_tail_rec_idx(arr: list) -> float:
    length = len(arr)

    def helper(idx: int, acc=0):
        if idx >= length:
            return acc/length
        # Also not using slices, saves memory
        return helper(idx+1, acc + arr[idx])

    return helper(0)


def array_mean_gen(arr: list) -> float:
    # No need to count length since we pass it in the first call to a helper func
    length = len(arr)

    def helper(idx: int, acc=0):
        if idx >= length:
            yield acc/length
        # Also not using slices, saves memory
        yield helper(idx+1, acc + arr[idx])

    yield helper(0)


arr = [i for i in range(1, 11)]
print(array_mean(arr))
print(array_mean_rec(arr))
print(array_mean_tail_rec(arr))
print(array_mean_tail_rec_idx(arr))
# print(array_mean_rec([i for i in range(10_000)]))
print(tramp(array_mean_gen, [i for i in range(1, 10_001)]))

print(list(itertools.accumulate(arr, lambda x, y: x+y))[len(arr)-1]/len(arr))
print(functools.reduce(lambda x, y: x+y, arr)/len(arr))

File: typescript/src/array_mean.ts
// Mean of Array using Iteration and Recursion
import { Trampoline, tramp } from "./global";

function arrayMean(arr: number[]): number {
    let sum = 0;
    const length = arr.length;
    arr.forEach((val) => {
        sum += val;
    })
    return sum / length;
}

function arrayMeanRec(arr: number[]): number {
    const length = arr.length;
    function helper(idx: number): any {
        if (idx === length) {
            return 0
        }
        const curr = arr[idx];
        return curr + helper(idx + 1);
    }
    return helper(0) / length;
}

function arrayMeanTailRec(arr: number[]): number {
    const length = arr.length;
    function helper(idx: number, acc = 0): any {
        if (idx === length) {
            return acc / length;
        }
        const curr = arr[idx];
        return helper(idx + 1, acc + curr);
    }
    return helper(0);
}

function arrayMeanThunk(arr: number[]): Trampoline<number> {
    const length = arr.length;
    function helper(idx: number, acc = 0): Trampoline<number> {
        if (idx === length) {
            return acc / length;
        }
        const curr = arr[idx];
        return () => helper(idx + 1, acc + curr);
    }
    return helper(0);
}

const arr = Array.from({ length: 10 }, (_, i) => i + 1);
const long_arr = Array.from({ length: 10_000 }, (_, i) => i + 1);
console.log(arrayMean(arr));
console.log(arrayMeanRec(arr));
console.log(arrayMeanTailRec(arr));
//console.log(arrayMeanTailRec(long_arr));
console.log(tramp(arrayMeanThunk, long_arr));